Library/Homebrew/utils/io_selector.rb

```diff
@@ -0,0 +1,67 @@
+require "delegate"
+
+require "extend/io"
+
+module Utils
+  class IOSelector < DelegateClass(Hash)
+    attr_reader :separator
+
+    alias all_streams keys
+    alias all_tags values
+    alias tag_of fetch
+
+    def self.each_line_from(streams = {}, separator = $/, &b)
+      new(streams, separator).each_line_nonblock(&b)
+    end
+
+    def each_line_nonblock(&b)
+      each_line_nonblock_enum.each(&b)
+    end
+
+    def pending_streams
+      @pending_streams ||= all_streams.dup
+    end
+
+    def initialize(streams = {}, separator = $/)
+      super(streams.invert.compare_by_identity)
+      @separator = separator
+    end
+
+    private
+
+    def each_line_nonblock_enum
+      Enumerator.new do |yielder|
```

Comment:
I didn't think about calling it without a block, since the only current use for this is with a block. This is neat, but I'm not sure if there will be a use for this without a block.